src/core.jl

```diff
@@ -175,7 +175,14 @@ output and `ȳ` the reverse-mode sensitivity of `y`.
 ∇(y::Node, ȳ) = propagate(tape(y), reverse_tape(y, ȳ))
 @inline ∇(y::Node{<:∇Scalar}) = ∇(y, one(unbox(y)))
 
-@inline ∇(x̄, f, ::Type{Arg{N}}, args...) where N = x̄ + ∇(f, Arg{N}, args...)
+# This is a fallback method where we don't necessarily know what we'll be adding and whether
+# we can update the value in-place, so we'll try to be clever and dispatch.
+@inline ∇(x̄, f, ::Type{Arg{N}}, args...) where {N} = update!(x̄, ∇(f, Arg{N}, args...))
+# Use broadcast for mixed array/scalar operations
+@inline update!(x̄::∇Scalar, y::∇Array) = x̄ .+ y
+@inline update!(x̄::∇Array, y::∇ArrayOrScalar) = (x̄ .= x̄ .+ y)
```

Comment:
```suggestion
@inline update!(x̄::∇Array, y::∇ArrayOrScalar) = x̄ .+= y
```

Reply:
For some reason I thought that was not in place, but it is.